featuretools/computational_backends/pandas_backend.py

```diff
@@ -229,15 +171,190 @@ def calculate_all_features(self, instance_ids, time_last,
+                also linked to all ancestors.
+
+            parent_df (pd.DataFrame): The data for the parent entity. Only
+                required if ancestor_relationship_variables is present.
+        """
+
+        # Step 1: Get a dataframe for the given entity, filtered by the given
+        # conditions.
+
+        features = [self.feature_set.features_by_name[fname] for fname in feature_trie[[]]]
+        need_all_rows = any(f.primitive.uses_full_entity for f in features)
+        if need_all_rows:
+            query_variable = None
+            query_values = None
+        else:
+            query_variable = filter_variable
+            query_values = filter_values
+
+        entity = self.entityset[entity_id]
+        columns = _necessary_columns(entity, features)
+        df = entity.query_by_values(query_values,
+                                    variable_id=query_variable,
+                                    columns=columns,
+                                    time_last=self.time_last,
+                                    training_window=self.training_window)
+
+        # Step 2: Add variables to the dataframe linking it to all ancestors.
+
+        new_ancestor_relationship_variables = []
+        if parent_relationship:
+            if ancestor_relationship_variables:
+                assert parent_df is not None
+
+                df, new_ancestor_relationship_variables = self._add_ancestor_relationship_variables(
+                    df, parent_df, ancestor_relationship_variables, parent_relationship)
+
+            # Add the variable linking this entity to its parent, so that
+            # descendants get linked to the parent.
+            new_ancestor_relationship_variables.append(parent_relationship.child_variable.id)
+
+        # Step 3: Recurse on children.
+
+        for edge, sub_trie in feature_trie.children():
+            is_forward, relationship = edge
+            if is_forward:
+                sub_entity = relationship.parent_entity.id
+                sub_filter_variable = relationship.parent_variable.id
+                sub_filter_values = df[relationship.child_variable.id]
+            else:
+                sub_entity = relationship.child_entity.id
+                sub_filter_variable = relationship.child_variable.id
+                sub_filter_values = df[relationship.parent_variable.id]
+
+            sub_df_trie = df_trie.get_node([edge])
+            self._calculate_features_for_entity(
+                entity_id=sub_entity,
+                feature_trie=sub_trie,
+                df_trie=sub_df_trie,
+                filter_variable=sub_filter_variable,
+                filter_values=sub_filter_values,
+                parent_relationship=(not is_forward) and relationship,
+                ancestor_relationship_variables=(not is_forward) and new_ancestor_relationship_variables,
+                parent_df=(not is_forward) and df)
```

Comment:
Since the above three params are all used for creating the relationship variables, and are all only necessary when the path is backward, I have been considering passing them as a single `parent_data` tuple.

Reply:
makes sense